sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java

```diff
@@ -0,0 +1,494 @@
+   */
+  void start(Iterable<ShardCheckpoint> checkpoints) {
+    LOG.info(
+        "Pool {} - starting for stream {} consumer {}. Checkpoints = {}",
+        poolId,
+        read.getStreamName(),
+        consumerArn,
+        checkpoints);
+    for (ShardCheckpoint shardCheckpoint : checkpoints) {
+      checkState(
+          !state.containsKey(shardCheckpoint.getShardId()),
+          "Duplicate shard id %s",
+          shardCheckpoint.getShardId());
+      ShardState shardState =
+          new ShardState(
+              initShardSubscriber(shardCheckpoint), shardCheckpoint, watermarkPolicyFactory);
+      state.put(shardCheckpoint.getShardId(), shardState);
+    }
+  }
+
+  /**
+   * Returns the next disaggregated {@link KinesisRecord} if available and updates {@link #state}
+   * accordingly so that it reflects a mutable checkpoint AFTER returning that record.
+   *
+   * <p>Async subscription errors are delayed until {@link #eventQueue} is completely drained and
+   * then rethrown here.
+   *
+   * <p>This repeats the following steps until a record or {@code null} was returned:
+   *
+   * <ol>
+   *   <li>If {@link #current} is null and {@link #eventQueue} is empty, return {@code null} unless
+   *       {@link #subscriptionError} is set: in that case rethrow.
+   *   <li>Otherwise if {@link #current} is null, poll next from {@link #eventQueue}.
+   *   <li>If {@link #current} has a next {@link KinesisClientRecord}, update {@link #state}
+   *       accordingly and return the corresponding converted {@link KinesisRecord}, optionally
+   *       triggering {@link #onEventDone} if that was the last record of {@link #current}.
+   *   <li>Finally, if nothing was returned yet, trigger {@link #onEventDone} and continue loop.
+   * </ol>
+   *
+   * <p>It polls the {@link #eventQueue} in a while loop to avoid returning null immediately if an
+   * event without records arrived. There may be events with records after the {@link #current}, and
+   * it is better to poll again instead of having {@link EFOKinesisReader#advance()} signalling
+   * false to Beam. Otherwise, Beam would poll again later, which would introduce unnecessary delay.
+   */
+  @Nullable
+  KinesisRecord getNextRecord() throws IOException {
+    while (true) {
+      if (!isStopped && subscriptionError != null) {
+        // Stop the pool to cancel all subscribers and prevent new subscriptions.
+        // Doing this as part of getNextRecord() avoids concurrent access to the state map and
+        // prevents any related issues.
+        stop();
+      }
+
+      if (current == null) {
+        current = eventQueue.poll();
+      }
+
+      if (current != null) {
+        String shardId = current.shardId;
+        ShardState shardState = Preconditions.checkStateNotNull(state.get(shardId));
+        if (current.hasNext()) {
+          KinesisClientRecord r = current.next();
+          KinesisRecord kinesisRecord = new KinesisRecord(r, read.getStreamName(), shardId);
```

Comment:
> I think, at this point, this is the only doubt I actually have. Other things are already addressed, right @mosche ?

@psolomin yes, this is the only open issue I'm aware of. Before returning, `shardState` must be adequately updated if this was the last record of `current` to make sure checkpoint marks are correctly reflecting progress.

```
          KinesisClientRecord r = current.next();
          // Make sure to update shard state accordingly if `current` does not contain any more
          // events. This is necessary to account for any re-sharding, so we could correctly resume
          // from a checkpoint if taken once we advanced to the record returned by getNextRecord().
          if (!current.hasNext()) {
            onEventDone(shardState, current);
            current = null;
          }
          KinesisRecord kinesisRecord = new KinesisRecord(r, read.getStreamName(), shardId);
          if (shardState.isAfterInitialCheckpoint(kinesisRecord)) {
            shardState.update(kinesisRecord);
            return kinesisRecord;
          }
```

Looks like we're missing a test case for this as follows:
- stub `subscribeToShard` with a reshard event that **also** contains 1 record.
- wait for exactly that 1 record
- verify the checkpoint mark now contains the new shards, but not the old one anymore

Reply:
> reshard event that also contains 1 record

I am not sure such a thing exists. Re-shard events always come with `continuationSequenceNumber = null`, and they never carry records with non-null sequence numbers.

Reply:
> I was convinced that re-shard events always come with continuationSequenceNumber = null

`continuationSequenceNumber` is unrelated to this. If it is `null` it simply means the shard doesn't contain any next records. But that doesn't mean the current event did not contain records.

Have you seen any documentation that explicitly states that events containing child shards do never contain records? If not we have to assume that this could be the case. Even if that is something we've never observed.

Reply:
> Even if that is something we've never observed.

Ok, got it. I am fine to go defensive here and assume that `SubscribeToShardEvent` can carry records AND child shards together. Let me try to put this out in the code, there might be other places it will be a trouble - I implicitly assumed this never happens.

Reply:
> Looks like we're missing a test case

Added it.